File: JavaSandbox/src/leetcode/lc394_DecodeString.java
// https://leetcode.com/problems/decode-string/
package leetcode;

public class lc394_DecodeString {
    public String decodeString(String s) {
        ArrayList<Character> stack = new ArrayList<>();
        
        for(int i=0; i < s.length(); i++) {
            Character ch = s.charAt(i);
            stack.add(ch);
            
            //System.out.println("stack A: " + stack);
            if(stack.get(stack.size()-1).equals(']')) {
                stack.remove(stack.size()-1); //pop off closing bracket
                String backtrackedStr = backtrackStr(stack);
                backtrackedStr = new StringBuilder(backtrackedStr).reverse().toString();
                //System.out.println("backtrackedStr: " + backtrackedStr);
                stack.remove(stack.size()-1); //pop off opening bracket
                //System.out.println("stack B: " + stack);
                
                // get k
                int k = 1;
                String kStr = "";
                while(!stack.isEmpty() && Character.isDigit(stack.get(stack.size()-1))) {
                    kStr = stack.remove(stack.size()-1).toString() + kStr;
                }
                if(!kStr.isEmpty())
                    k = Integer.parseInt(kStr);
                
                String section = backtrackedStr.repeat(k);
                //System.out.println("section: " + section);
                
                //put section back on stack
                for(int j=0; j < section.length(); j++) {
                    stack.add(section.charAt(j));
                }
                //System.out.println("stack end of iter: " + stack);
            } 
        }
        
        String sdecoded = "";
        while(!stack.isEmpty()) {
            sdecoded += stack.remove(0);
        }
        return sdecoded;
    }
    
    public String backtrackStr(ArrayList<Character> stack) {
        String s = "";
        while(!stack.isEmpty() && !(stack.get(stack.size()-1).equals('['))) {
            Character ch = stack.remove(stack.size()-1);
            s = s.concat(ch.toString());
        }
        
        return s;
    }
}

File: PythonSandbox/src/leetcode/lc394_decode_string.py
'''
https://leetcode.com/problems/decode-string/
'''
class Solution:
    def decodeString(self, s: str) -> str:
        if not s:
            return ""
        
        stack = [s[0]]
        dcStr = ""
        i = 1
        while i < len(s):
            ch = s[i]
            stack.append(ch)
            
            if stack[-1] == "]":
                stack.pop() # pop off closing bracket
                backtrackedStr, stack = self.goToOpenBracket(stack)
                
                k = ""
                if stack:
                    stack.pop() # pop off opening bracket
                    while stack and stack[-1].isnumeric():
                        k = stack.pop() + k # pop off k
                    k = int(k)
                else:
                    k = 1
                    
                #print("stack B: ", stack)
                decodedSection = k * backtrackedStr[::-1]
                #print("dcStr B: ", dcStr)
                for ch in decodedSection:
                    stack.append(ch)
                #print("stack after adding back: ", stack)
            
            i += 1 
                
        return "".join(stack)
            
    def goToOpenBracket(self, stack):
        string = ""
        while stack and stack[-1] != "[":
            string += stack.pop()
        return string, stack